harness/determined/profiler.py

```diff
@@ -105,29 +155,45 @@ def update_batch_idx(self, new_batch_idx: int):
                 self._end_collection()
                 self.has_finished = True
 
+    def __enter__(self) -> 'SystemProfilerAgent':
+        return self
+
+    def __exit__(self):
+        self._end_collection()
 
     def _begin_collection(self):
         self.sys_metric_collector_thread.activate()
         # TODO: Start up TimingBatcher as well
 
     def _end_collection(self):
+        if self.killed:
+            return
+        self.killed = True
         self.sys_metric_collector_thread.kill()
         self.sender_thread.kill()
         # TODO: Shut down TimingBatcher as well
 
         self.sys_metric_collector_thread.join()
         self.sender_thread.join()
 
-
-    def record_timing(self):
+    @contextmanager
+    def record_timing(self, metric_name: str) -> None:
+        # TODO(brad): Check within batch range.
         if not self.is_enabled:
             return
-        # TODO: Add new timing to TimingBatcher
-
+        start = default_timer()
+        yield
```

Comment:
I don't think we should make the main public facing API a context manager. It's not very flexible for non-PyTorch frameworks.